ecllibrary/std/system/Store.ecl

```diff
@@ -0,0 +1,573 @@
+/*##############################################################################
+## HPCC SYSTEMS software Copyright (C) 2019 HPCC Systems®.  All rights reserved.
+############################################################################## */
+
+IMPORT Std;
+
+/**
+ * A specific instance of this module needs connection and user credential
+ * information, provided as arguments.
+ *
+ * @param   username        The username of the user requesting access to the
+ *                          key value store; set to an empty string if
+ *                          authentication is not required; OPTIONAL, defaults
+ *                          to an empty string
+ * @param   userPW          The password of the user requesting access to the
+ *                          key value store; set to an empty string if
+ *                          authentication is not required; OPTIONAL, defaults
+ *                          to an empty string
+ * @param   espURL          The full URL for accessing the esp process running
+ *                          on the HPCC Systems cluster (this is typically the
+ *                          same URL as used to access ECL Watch); set to an
+ *                          empty string to use the URL of the current esp
+ *                          process; OPTIONAL, defaults to an empty string
+ *
+ * @return  A reference to the module, correctly initialized with the
+ *          given access parameters.
+ */
+EXPORT Store(STRING username = '',
+             STRING userPW = '',
+             STRING espURL = '') := MODULE
+
+    SHARED ENCODED_CREDENTIALS := IF
+        (
+            TRIM(username, ALL) != '',
```

Comment:
Not if a user submits ' ' -- a string composed of spaces -- as a username.  It is just a bit of self-protection.  Is it likely to produce performance issues?

Granted, trimming could be removed and trim issues with the username and password could be handled via rejection from the esp service.  Would that be a better option?

Reply:
All strings are effectively right trimmed before being compared (trailing spaces are ignored).  So comparing with a blank string is identical to TRIM(username) = '' or TRIM,ALL.
In this context it is almost certainly going to make no difference at all.  If the trim was applied for each input record being processed in a large dataset then it would.